examples/repo_lock.py

```diff
@@ -0,0 +1,74 @@
+from pubtools.pulplib import Client
```

Comment:
Can you please fix up some consistency issues between this example and the others?

- naming: `_` vs `-`
- naming: `.py` vs no extension
- should be marked executable and start with a shebang line

Reply:
Hey Rohan, do you have any general advice for how to "make it work" regarding the FakeClient? 

The request function lambda requires three attributes that don't exist in the FakeClient class. A compounding issue is that one of these attributes then tries to get the method. Is this just a case of yes everything there will need to be set up? 

Looking through the repository code I can't find any special considerations for the FakeClient class, so I assume it's a case of modifying the FakeClient class to get it to work with the rest of the code as is.

Reply:
Firstly, though I said just "don't crash" could be the goal before, later I thought ahead to the changes needed in pubtools-pulp-maintenance-on to make use of this new feature and how it would be covered in unit tests there.

In fact it would be quite valuable if there were a way, when using the fake client, to access the repo lock history (same as today you can access the repo publish history).  The way of testing the pubtools-pulp-maintenance-on change would then be as easy as, in existing test_maintenance_on test there, something like:

```
# It should have locked and then unlocked the repo
assert len(controller.repository_lock_history) == 2
assert controller.repository_lock_history[0].repository.id == "redhat-maintenance"
assert controller.repository_lock_history[0].event_type == "lock"
assert controller.repository_lock_history[1].repository.id == "redhat-maintenance"
assert controller.repository_lock_history[1].event_type == "unlock"
```

So I think making something like the above work should be the goal.

The usual approach so far has been for the real and fake clients to implement some of the same (private) API.  For example, they both implement a `def _publish_repository` which takes care of real or fake publishing.  A fair bit of refactoring would be needed here to follow that approach though.

I wouldn't expect this to be handled at the HTTP request layer, i.e. adding `_request_executor` and `_do_request` onto the fake client wouldn't be a good approach I think.

I think there are two reasonable approaches to take here:

1) refactor it so it works similarly to the current approach already used for repo publishing (`def _publish_repository` and file uploads (`_do_upload`). Give real and fake clients different implementations of a compatible API for locking/unlocking repos, then use it from `RepoLock` class.  Real client does the same kind of HTTP requests as currently prepared in `RepoLock`, fake client always succeeds (as long as repo exists) and just stores a lock/unlock event on a new attribute in FakeState class.

2) refactor so that `RepoLock` receives a reference to a client object.  Then it can explicitly check if the client object is a fake client and, if so, it can jump to a different simplified implementation.